### PR TITLE
Remove caching from `OpaqueTypeExpander`. Fixes #87450.

### LINKED ISSUE #87450: Exponential compile time with recursive opaque type
### Code

I tried this code:

```rust
fn bar() -> impl Fn() {
    wrap(wrap(wrap(wrap(foo()))))
}

fn foo() -> impl Fn() {
    wrap(wrap(wrap(wrap(wrap(wrap(wrap(foo())))))))
}

fn wrap(f: impl Fn()) -> impl Fn() {
    move || f()
}
```

I expected running `cargo check` to produce an error like:
```
error[E0720]: cannot resolve opaque type
 --> src/lib.rs:5:13
  |
5 | fn foo() -> impl Fn() {
  |             ^^^^^^^^^ recursive opaque type
6 |     ((wrap(wrap(wrap(wrap(wrap(foo())))))))
  |     --------------------------------------- returning here with type `impl Fn<()>`
...
9 | fn wrap(f: impl Fn()) -> impl Fn() {
  |                          --------- returning this opaque type `impl Fn<()>`
```

It does, but `cargo check` takes time exponential in the number of `wrap` calls, although doesn't appear to use any extra memory.
In `foo` the time take looks like:
| Num `wraps`s | Time taken (s) |
|-|-|
| 4 | 0.5 |
| 5 | 2.1 |
| 6 | 28 |
| 7 | 436 |

I believe it's also exponential in the number of `wrap` calls in `bar`.

The reason I had code like this was because I was using nom's parser combinators which results in lots of nested closures. I spotted it because this caused rust-analyzer to hang because cargo clippy hung (it looks like check was the issue), I had to kill clippy to fix it.

### Bisection results
I bisected this (with cargo-bisect-rustc) to find the regression in `nightly-2019-10-16`, which contained the following bors commits:
1. Auto merge of #65433 - Centril:rollup-rzvry15, r=Centril
1. Auto merge of #65454 - tmandry:rollup-0k6jiik, r=tmandry
1. Auto merge of #65450 - Manishearth:clippyup, r=Manishearth
1. Auto merge of #65172 - tanriol:explain_borrow-use-context-dominators, r=nagisa
1. Auto merge of #65445 - ehuss:update-cargo-books, r=alexcrichton

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged

### PR DESCRIPTION
**Pull request by hkratz**
Draft for discussion and to run tests and timing tests.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/util.rs [hunk 1]
```diff
@@ -557,7 +556,6 @@ struct OpaqueTypeExpander<'tcx> {
     seen_opaque_tys: FxHashSet<DefId>,
     // Cache of all expansions we've seen so far. This is a critical
     // optimization for some large types produced by async fn trees.
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/util.rs:558]:
```suggestion
```

**Comment by jyn514** [comment]
@rust-timer queue

**Comment by jyn514** [comment]
Oh, finally found it.

@rust-timer build 8c29eb63f4dc13a3a8c06c8eacc2b437574158ec

**Comment by hkratz** [comment]
That would have been too easy.

**Pull request closed without merging**